### ISSUE TITLE
Hygienic quasiquote gives NameError

### ISSUE THREAD

**Issue by thirtythreeforty** [Technologicat/mcpyrate#21]
With the following macro definition:

```
def mymacro(tree, **kw):
    foobar = 0
    return q[h[foobar]]
```
I get a NameError complaining about `mcpyrate`:
```
>       x = mymacro[3]
E       NameError: name 'mcpyrate' is not defined
```

This is true - at the macro invocation site, I have not imported any mcpyrate modules.

This seems to be because `h[]` finally expands (unhygienically :) to a call to `mcpyrate.quotes.lookup_value`, which is not in scope.
```
collecting ... **Tree 0x7f4637e6cf50 (/home/georgev/Code/nmigen-decorator/sonata-gateware/tests/impl_genlogic.py) before macro expansion:
  L   12 x = mymacro[3]
**Tree 0x7f4637e6cf50 (/home/georgev/Code/nmigen-decorator/sonata-gateware/tests/impl_genlogic.py) after step 1:
  L   12 x = mcpyrate.quotes.lookup_value(('foobar', b'\x80\x03K\x00.'))
**Tree 0x7f4637e6cf50 (/home/georgev/Code/nmigen-decorator/sonata-gateware/tests/impl_genlogic.py) macro expansion complete after 1 step.
collected 2 items                                                                                                                                    
```

**Comment by thirtythreeforty** [comment]
A thought: is there a reason the lookup needs to occur in a library function? Could the pickled value be emitted directly into the desugared code?

**Comment by Technologicat** [comment]
Aaaaa, silly of me not to have noticed this. Thanks for the catch! Your analysis of the reason of the crash is correct.

The thinking behind why I made the expansion into a call to a library function was to make *hygienically unquoted value* into a SICP-style *data abstraction*. That is to say, the use site doesn't need to care about the implementation details of how that value is internally represented, as long as there's an API function that takes the opaque data and decodes it. That's better for readability, since we can immediately see which subsystem that call belongs to. At the same time, I wanted the "opaque" data to be as transparent as possible, to facilitate debugging crashes like this.

It's a good point to ask whether we really need that library function.

We cannot desugar directly into a pickled value, because a pickled value by itself is essentially just a binary dump of attributes plus possibly a fully qualified class name to load that data into - it must be unpickled before it can be used. Maybe it should be mentioned here that beside that binary data, the unpickler needs the original, or at least data-compatible, class definition (of the value's type) to be loadable via the same dotted name it had when the value was pickled. I suppose the Python devs designed it like this to maximize flexibility, by providing the `__setstate__` method hook with which we can override the default unpickling behavior for a given class.

On the other hand, we cannot desugar into a direct reference to a live value, due to the need to support bytecode caching. I initially experimented with a design that simply stored the hygienically unquoted values as uniquely keyed object references in a dictionary (at macro expansion time), without pickling, but there was the problem that when the program starts again, if macro expansion time was last week, then (because the macro-expanded code is being loaded from the bytecode cache) the dictionary will be empty. :)

(This is how I discovered why `macropy` uses `pickle` in its hygienic quasiquote system. It's needed to store the values across program restarts, since macro expansion occurs only during the first run.)

So we need at least pickle, and we must be able to call `pickle.loads` to perform the unpickling... and that function must be imported, so it's no better than `mcpyrate.quotes.lookup_value`.

Also, `mcpyrate.quotes.lookup_value` has an important responsibility - it implements the "unpickle only once per use site per Python process" feature of hygienically unquoted values.

(Now that I look at the source code, though, I notice there may be a bug regarding that behavior. Several use sites of `lookup_value` may produce the same value instance if the `name` part of the cache key is the same, and the bytestring containing the binary data happens to be interned by the Python interpreter. To fix that, I'll need to change the format of the cache key to include a UUID.)

So maybe we should keep the library function...

An idea, it might be possible to abuse the builtin `__import__` function to bring what we need into scope?

```python
__import__("mcpyrate.quotes", globals(), locals(), [], 0).quotes.lookup_value(...)
```

where the `...` are the arguments to `lookup_value`. The call to [`__import__`](https://docs.python.org/3/library/functions.html#__import__) returns the top-level module that was imported, in this case `mcpyrate`.

This is still a bit hacky, but should work. I'll have to test it.

Further ideas for better solutions are appreciated!

**Comment by thirtythreeforty** [comment]
The justification for keeping the library function makes sense.  So, then it just becomes a case of how to call it.

You're going to have to import it eventually, and you really only have two options that I see:
- use a builtin, such as __import__. Arguably this isn't a hack :)
- ask the macro author to provide a hook back that the emitted code can call back into. I'm currently doing this as a workaround by emitting an `import mcpyrate` statement in the same block that calls the hygienically bound state.

**Comment by Technologicat** [comment]
I did some quick testing and, although the docs warn against using it directly, it seems `__import__` does behave sensibly - if we use it to import the same module twice, it will re-use the result of the first import, just like the import statement does.

This was somewhat surprising, since a large part of Python's import mechanism is implemented in Python, in [`importlib/_bootstrap_external.py`](https://github.com/python/cpython/blob/master/Lib/importlib/_bootstrap_external.py). It seems the caching comes from a lower level in the interpreter, which is good for our purposes.

I'll fix this using the `__import__` builtin.

**Comment by Technologicat** [comment]
Should be fixed. Could you pull from git and test on 3.7 or 3.8?

**Comment by thirtythreeforty** [comment]
This works for me! Thank you!

**Comment by Technologicat** [comment]
Great! Thanks for confirming. The fix will be included in 3.1.0.
